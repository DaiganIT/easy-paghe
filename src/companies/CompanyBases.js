import React from 'react';
import classnames from 'classnames';
import {
	withStyles,
	TextField,
	Table,
	TableBody,
	TableRow,
	TableCell,
	Typography,
	IconButton
} from '@material-ui/core';
import { Delete, Add } from '@material-ui/icons';

const styles = theme => ({
	paper: {
		padding: '1em',
		marginBottom: '1em',
		position: 'relative'
	},
	miniTitle: {
		marginBottom: '1em',
	},
	hash: {
		textAlign: 'center',
	},
	fab: {
		marginBottom: '1em'
	},
	addBase: {
		cursor: 'pointer'
	},
	colPadding: {
		border: 0,
		padding: '10px'
	}
});

function CompanyBases({ classes, bases, addBase, deleteBase, isSaving, updateBaseField, errors }) {
	return (
		<React.Fragment>
			<Table>
				<TableBody>
					{bases.map((base, index) => (
						<TableRow key={base.id || index}>
							<TableCell className={classes.colPadding} >
								<TextField
									id={`company-base-${index}-name`}
									required
									error={!!errors.bases && !!errors.bases[index] && !!errors.bases[index].name}
									variant="outlined"
									label="Nome"
									fullWidth
									className={classes.textField}
									disabled={isSaving}
									value={base.name}
									onChange={(e) => updateBaseField('name', index, e.target.value)}
								/></TableCell>
							<TableCell className={classes.colPadding}>
								<TextField
									id={`company-base-${index}-address`}
									error={!!errors.bases && !!errors.bases[index] && !!errors.bases[index].address}
									variant="outlined"
									label="Indirizzo"
									fullWidth
									className={classes.textField}
									disabled={isSaving}
									value={base.address}
									onChange={(e) => updateBaseField('address', index, e.target.value)}
								/>
							</TableCell>
							<TableCell component="th" scope="row" padding="checkbox" className={classnames(classes.deleteClass, classes.colPadding)} align="right">
								{index > 0
									? <IconButton id={`company-base-${index}-delete`} onClick={() => deleteBase({ baseId: base.id, index })}>
										<Delete color="secondary" />
									</IconButton>
									: null}
							</TableCell>
						</TableRow>
					))}
					<TableRow hover>
						<TableCell component="th" scope="row" colSpan={3} onClick={addBase} align="center" className={classes.addBase}>
							<Typography id="add-company-button"><Add color="secondary" /> Aggiungi sede</Typography>
						</TableCell>
					</TableRow>
				</TableBody>
			</Table>
		</React.Fragment>
	);
}

export default withStyles(styles)(CompanyBases);
