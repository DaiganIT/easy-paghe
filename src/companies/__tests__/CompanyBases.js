import React from 'react';
import {
  render,
  cleanup,
  fireEvent,
} from 'react-testing-library';
import 'jest-dom/extend-expect';
import CompanyBases from '../CompanyBases';

afterEach(cleanup)

const defaultProps = {
  addBase: jest.fn(),
  deleteBase: jest.fn(),
  isSaving: false,
  bases: [
    {
      name: 'Main Base',
      address: 'The address'
    }
  ],
  updateBaseField: jest.fn(),
  errors: {}
}

test('Company Bases structure snapshot is correct', () => {
  const { container } = render(<CompanyBases {...defaultProps} />);
  expect(container.firstChild).toMatchSnapshot();
});

const testCases = [
  {
    name: 'name',
    inputId: 'name'
  },
  {
    name: 'address',
    inputId: 'address'
  },
]

testCases.forEach(testCase => {
  afterEach(jest.clearAllMocks);

  test(`Company Bases ${testCase.name} input is rendered`, () => {
    const { container } = render(<CompanyBases {...defaultProps} />);

    expect(container.querySelector(`input#company-base-0-${testCase.inputId}`).value).toBe(defaultProps.bases[0][testCase.name]);
  });

  test(`Company Bases ${testCase.name} input change is catched`, () => {
    const { container } = render(<CompanyBases {...defaultProps} />);

    fireEvent.change(container.querySelector(`input#company-base-0-${testCase.inputId}`), { target: { value: 'Changed' } });
    expect(defaultProps.updateBaseField).toHaveBeenCalledWith(testCase.name, 0, 'Changed');
  });
});

test('Company Details inputs are disabled when isSaving is true', () => {
  const props = Object.assign({}, defaultProps, { isSaving: true });
  const { container } = render(<CompanyBases {...props} />);

  const allInputs = container.querySelectorAll('input');
  for (const input of allInputs) {
    expect(input).toBeDisabled();
  }
});

const testCases2 = [
  {
    name: 'name',
    inputId: 'name'
  },
  {
    name: 'address',
    inputId: 'address'
  },
]

testCases2.forEach(testCase => {
  test(`Company Bases ${testCase.name} input is rendered with errors`, () => {
    const props = Object.assign({}, defaultProps, { errors: { bases: [{ [testCase.name]: ['error'] }] } });
    const { container } = render(<CompanyBases {...props} />);

    expect(container.querySelector(`label[for=company-base-0-${testCase.inputId}]`).className).toContain('MuiFormLabel-error');
  });
});

test('Company Bases first company base has no delete button', () => {
  const { container } = render(<CompanyBases {...defaultProps} />);

  const deleteButton = container.querySelector(`#company-base-0-delete`);
  expect(deleteButton).not.toBeInTheDocument();
});

test('Company Bases second company base has delete button', () => {
  const props = Object.assign({}, defaultProps, {
    bases: [
      { name: 'Main Base', address: 'The address' },
      { name: 'Main Base 2', address: 'The address 2' }
    ]
  });
  const { container } = render(<CompanyBases {...props} />);

  const deleteButton = container.querySelector(`#company-base-1-delete`);
  expect(deleteButton).toBeInTheDocument();
});

test('Company Bases first company base has no delete button', () => {
  const props = Object.assign({}, defaultProps, {
    bases: [
      { id: 5, name: 'Main Base', address: 'The address' },
      { id: 10, name: 'Main Base 2', address: 'The address 2' }
    ]
  });
  const { container } = render(<CompanyBases {...props} />);

  const deleteButton = container.querySelector(`#company-base-1-delete`);
  fireEvent.click(deleteButton);
  expect(props.deleteBase).toHaveBeenCalledWith({ baseId: 10, index: 1 });
});

test('Company Bases add new company button triggers the event', () => {
  const { container } = render(<CompanyBases {...defaultProps} />);

  const addButton = container.querySelector(`#add-company-button`);
  fireEvent.click(addButton);
  expect(defaultProps.addBase).toHaveBeenCalled();
});