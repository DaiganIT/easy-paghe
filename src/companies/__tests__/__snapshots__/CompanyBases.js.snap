// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company Bases structure snapshot is correct 1`] = `
<table
  class="MuiTable-root-7"
>
  <tbody
    class="MuiTableBody-root-8"
  >
    <tr
      class="MuiTableRow-root-9"
    >
      <td
        class="MuiTableCell-root-14 MuiTableCell-body-16 CompanyBases-colPadding-6"
      >
        <div
          class="MuiFormControl-root-26 MuiFormControl-fullWidth-29"
        >
          <label
            class="MuiFormLabel-root-41 MuiFormLabel-filled-45 MuiFormLabel-required-46 MuiInputLabel-required-34 MuiInputLabel-root-30 MuiInputLabel-formControl-35 MuiInputLabel-animated-38 MuiInputLabel-shrink-37 MuiInputLabel-outlined-40"
            data-shrink="true"
            for="company-base-0-name"
          >
            Nome
            <span
              class="MuiFormLabel-asterisk-47"
            >
               *
            </span>
          </label>
          <div
            class="MuiInputBase-root-61 MuiOutlinedInput-root-48 MuiInputBase-fullWidth-70 MuiInputBase-formControl-62"
          >
            <fieldset
              aria-hidden="true"
              class="MuiPrivateNotchedOutline-root-78 MuiOutlinedInput-notchedOutline-55"
              style="padding-left: 8px;"
            >
              <legend
                class="MuiPrivateNotchedOutline-legend-79"
                style="width: 0px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
            <input
              aria-invalid="false"
              class="MuiInputBase-input-71 MuiOutlinedInput-input-56"
              id="company-base-0-name"
              required=""
              type="text"
              value="Main Base"
            />
          </div>
        </div>
      </td>
      <td
        class="MuiTableCell-root-14 MuiTableCell-body-16 CompanyBases-colPadding-6"
      >
        <div
          class="MuiFormControl-root-26 MuiFormControl-fullWidth-29"
        >
          <label
            class="MuiFormLabel-root-41 MuiFormLabel-filled-45 MuiInputLabel-root-30 MuiInputLabel-formControl-35 MuiInputLabel-animated-38 MuiInputLabel-shrink-37 MuiInputLabel-outlined-40"
            data-shrink="true"
            for="company-base-0-address"
          >
            Indirizzo
          </label>
          <div
            class="MuiInputBase-root-61 MuiOutlinedInput-root-48 MuiInputBase-fullWidth-70 MuiInputBase-formControl-62"
          >
            <fieldset
              aria-hidden="true"
              class="MuiPrivateNotchedOutline-root-78 MuiOutlinedInput-notchedOutline-55"
              style="padding-left: 8px;"
            >
              <legend
                class="MuiPrivateNotchedOutline-legend-79"
                style="width: 0px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
            <input
              aria-invalid="false"
              class="MuiInputBase-input-71 MuiOutlinedInput-input-56"
              id="company-base-0-address"
              type="text"
              value="The address"
            />
          </div>
        </div>
      </td>
      <th
        class="MuiTableCell-root-14 MuiTableCell-body-16 MuiTableCell-alignRight-24 MuiTableCell-paddingCheckbox-20 CompanyBases-colPadding-6"
        scope="row"
      />
    </tr>
    <tr
      class="MuiTableRow-root-9 MuiTableRow-hover-11"
    >
      <th
        class="MuiTableCell-root-14 MuiTableCell-body-16 MuiTableCell-alignCenter-23 CompanyBases-addBase-5"
        colspan="3"
        scope="row"
      >
        <p
          class="MuiTypography-root-80 MuiTypography-body1-89"
          id="add-company-button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-116 MuiSvgIcon-colorSecondary-118"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
           Aggiungi sede
        </p>
      </th>
    </tr>
  </tbody>
</table>
`;
